### PR TITLE
[FLINK-17166][dist] Modify the log4j-console.properties to also output logs into the files for WebUI

### PR DESCRIPTION
**Pull request by wangyang0918**
## What is the purpose of the change
Currently, we when use `flink-console.sh` to start jobmanager/taskmanager start-foreground, the logs could not be accessed via webui. Just because we do not output the logs to the files.

In docker container environment, using `docker logs` to view the logs is a common way. So still need to output the logs to console.

We will add a `RollingFileAppender` to also output the logs to a local file so that the logs could be accessed via web dashboard.

## Brief change log

* Update `log4j-console.properties` and `logback-console.xml` to output logs to files
* Update `flink-console.sh`

## Verifying this change

* Current all e2e tests should not be affected
* Manually test logging/stdout/stderr could output to console and file at same time
  * `./bin/jobmanager.sh start-foreground`
  * `./bin/taskmanager.sh start-foreground`
  * View localhost:8081 to check the logs could be accessed via webui

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by wangyang0918** [comment]
cc @tillrohrmann @azagrebin Could you please have a look at your convenience?

**Comment by tillrohrmann** [comment]
Thanks for creating this PR @wangyang0918. Interesting solution! We'll take a closer look soon.

File: flink-runtime/src/test/java/org/apache/flink/runtime/util/StdOutErrRedirectorTest.java [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+
+		assertFalse(StdOutErrRedirector.isRedirecting());
+
+		proxyStream.print(intMsg);
+		assertThat(StdOutErrRedirector.isRedirecting(), is(false));
+		proxyStream.println(strMsg);
+		assertThat(StdOutErrRedirector.isRedirecting(), is(false));
+
+		assertThat(logContents.size(), is(3));
+		assertThat(logContents, Matchers.containsInAnyOrder(String.valueOf(intMsg), strMsg, System.lineSeparator()));
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/util/StdOutErrRedirectorTest.java:86]: the order should be important

File: flink-runtime/src/test/java/org/apache/flink/runtime/util/StdOutErrRedirectorTest.java [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+	@Test
+	public void testRecursivePrint() throws Exception {
+		final String recursive = "recursive";
+		final Logger logger = new TestingLogger();
+
+		final PrintStream proxyStream = StdOutErrRedirector.createLoggerProxy(logger, originalStream);
+		((TestingLogger) logger).setMsgConsumer(
+			msg -> {
+				assertThat(StdOutErrRedirector.isRedirecting(), is(true));
+				proxyStream.print(recursive);
```
Comment (zentol) [inline flink-runtime/src/test/java/org/apache/flink/runtime/util/StdOutErrRedirectorTest.java:98]:
```suggestion
				proxyStream.print(msg);
```

**Comment by wangyang0918** [comment]
@zentol Thanks a lot for your review.

I agree with you that introducing the stream redirections may take some problems. So we need to be very careful to enable the redirection via `StdOutErrRedirector.redirectStdOutErr`. Currently, i only see the requirements when we want to output the stdout/err to log files and console at the same time. Also i do not have seen a better solution. Using the `tee` will make us have to use the `grep` to filter out the log4j loggings. Just as the discussion in the [ticket](https://issues.apache.org/jira/browse/FLINK-17166?focusedCommentId=17087643&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17087643).

Also not all entrypoints need to do this. I think only the docker environment deployment need(e.g. standalone, K8s). We could put it in the first line in the main of `XXEntrypoint` and i do not find a logging system cache problem here.

The most important concerns when i have to introduce the stream redirection is not about the mixing the `.out` and `.err` file. It is we have to filter out the log4j logging from the stdout/err when using `tee`.
Actually, the YARN and K8s deployment always put the stdout to `.out` file and stderr to `.err` file. And i think it is better to separate them.

File: flink-dist/src/main/flink-bin/bin/flink-console.sh [hunk 3]
```diff
@@ -58,7 +58,18 @@ esac
...
+if [ "$FLINK_IDENT_STRING" = "" ]; then
+    FLINK_IDENT_STRING="$USER"
+fi
+
+RANDOM_ID=$(echo $RANDOM | tr '[0-9]' '[a-z]')
+
+FLINK_LOG_PREFIX="${FLINK_LOG_DIR}/flink-${FLINK_IDENT_STRING}-${SERVICE}-${RANDOM_ID}-${HOSTNAME}"
+log="${FLINK_LOG_PREFIX}.log"
+out="${FLINK_LOG_PREFIX}.out"
+err="${FLINK_LOG_PREFIX}.err"
```
[the remark this answers is not in the document]
Comment (wangyang0918) [reply flink-dist/src/main/flink-bin/bin/flink-console.sh:70]: Actually, the YARN and K8s deployment always put the stdout to `.out` file and stderr to `.err` file. And i think it is better to separate them if we could.

File: flink-dist/src/main/flink-bin/bin/flink-console.sh [hunk 4]
```diff
@@ -58,7 +58,18 @@ esac
...
 FLINK_TM_CLASSPATH=`constructFlinkClassPath`
 
-log_setting=("-Dlog4j.configuration=file:${FLINK_CONF_DIR}/log4j-console.properties" "-Dlog4j.configurationFile=file:${FLINK_CONF_DIR}/log4j-console.properties" "-Dlogback.configurationFile=file:${FLINK_CONF_DIR}/logback-console.xml")
+if [ "$FLINK_IDENT_STRING" = "" ]; then
+    FLINK_IDENT_STRING="$USER"
+fi
+
+RANDOM_ID=$(echo $RANDOM | tr '[0-9]' '[a-z]')
+
+FLINK_LOG_PREFIX="${FLINK_LOG_DIR}/flink-${FLINK_IDENT_STRING}-${SERVICE}-${RANDOM_ID}-${HOSTNAME}"
```
[the remark this answers is not in the document]
Comment (wangyang0918) [reply flink-dist/src/main/flink-bin/bin/flink-console.sh:67]: I think the foreground-start is a little different from daemon-start. We could not get the `pid` of jobmanager/taskmanager process. And use the same logic to generate the log file name. If you think it is necessary to use an `incremental` id, i could update this part.

File: flink-runtime/src/main/java/org/apache/flink/runtime/util/StdOutErrRedirector.java [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+
+	/**
+	 * Create logger proxy print stream.
+	 * Do not check null, keep the same behavior with System.out/err.
+	 *
+	 * @param originalPrintStream the original print stream
+	 * @return the proxy print stream
+	 */
+	@VisibleForTesting
+	static PrintStream createLoggerProxy(final Logger logger, final PrintStream originalPrintStream) {
```
[the remark this answers is not in the document]
Comment (wangyang0918) [reply flink-runtime/src/main/java/org/apache/flink/runtime/util/StdOutErrRedirector.java:65]: Nice suggestion. I will fix it.

**Comment by zentol** [comment]
First up, please remember to push updates to a PR as separate commits, so I can see easily which comments have been addressed.

>  Using the tee will make us have to use the grep to filter out the log4j loggings. Just as the discussion in the ticket.

As I said:

> As I understand it the only problem we had with the tee approach was that the logging output also ends up in .out .
I'm wondering if this is really such a big problem; if we define the .out output as what would be printed to the console, then this is perfectly acceptable behavior.

Moving on:

> Not all entrypoints need to do redirect the streams. Currently, i think only the docker environment deployment need(e.g. standalone, K8s).

This both implies different behaviors depending on how you deploy Flink, and can introduce subtle issues if the supported entry points change. This is exactly the kind of inconsistent behaviors we shouldn't be introducing.

> As you have said that the logging framework will cache the OutputStream at creation time, could you share me more insights here?

Unfortunately no, but I have been bitten by this in the past.

File: flink-dist/src/main/flink-bin/bin/flink-console.sh [hunk 6]
```diff
@@ -58,7 +58,19 @@ esac
...
 
-log_setting=("-Dlog4j.configuration=file:${FLINK_CONF_DIR}/log4j-console.properties" "-Dlog4j.configurationFile=file:${FLINK_CONF_DIR}/log4j-console.properties" "-Dlogback.configurationFile=file:${FLINK_CONF_DIR}/logback-console.xml")
+if [ "$FLINK_IDENT_STRING" = "" ]; then
+    FLINK_IDENT_STRING="$USER"
+fi
+
+pid=$FLINK_PID_DIR/flink-foreground-$FLINK_IDENT_STRING-$SERVICE.pid
+id=$(getFlinkDaemonID $pid 0)
+
+FLINK_LOG_PREFIX="${FLINK_LOG_DIR}/flink-foreground-${FLINK_IDENT_STRING}-${SERVICE}-${id}-${HOSTNAME}"
```
Comment (zentol) [inline flink-dist/src/main/flink-bin/bin/flink-console.sh:68]: I see no reason why we should have this in the name.
```suggestion
FLINK_LOG_PREFIX="${FLINK_LOG_DIR}/flink-${FLINK_IDENT_STRING}-${SERVICE}-${id}-${HOSTNAME}"
```

File: flink-dist/src/main/flink-bin/bin/config.sh [hunk 7]
```diff
@@ -660,3 +660,39 @@ extractExecutionParams() {
...
+    # The lock is created on the PID directory since a lock file cannot be safely
+    # removed. The daemon is started with the lock closed and the lock remains
+    # active in this script until the script exits.
+    command -v flock >/dev/null 2>&1
+    if [[ $? -eq 0 ]]; then
+        exec 200<"$FLINK_PID_DIR"
+        flock 200
+    fi
+
+    # Remove the pid file when all the processes are dead
```
Comment (zentol) [inline flink-dist/src/main/flink-bin/bin/config.sh:682]: this surely is not a side-effect this method should have

File: flink-dist/src/main/flink-bin/bin/config.sh [hunk 8]
```diff
@@ -660,3 +660,39 @@ extractExecutionParams() {
...
+                all_dead=1
+            fi
+        done < "$pid"
+        [ $clean_up_when_all_processes_dead -eq 0 -a $all_dead -eq 0 ] && rm $pid
+    fi
+
+    # Ascending ID depending on number of lines in pid file.
+    # This allows us to start multiple daemon of each type.
+    local id=$([ -f "$pid" ] && echo $(wc -l < "$pid") || echo "0")
+    echo $id
```
Comment (zentol) [inline flink-dist/src/main/flink-bin/bin/config.sh:697]: Can we not simplify this?
```suggestion
    echo $([ -f "$pid" ] && echo $(wc -l < "$pid") || echo "0")
```

File: flink-dist/src/main/flink-bin/bin/flink-console.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (zentol) [reply flink-dist/src/main/flink-bin/bin/flink-console.sh:71]: > And i think it is better to separate them if we could.

That may be so, but the scripts should have a consistent behavior.

**Comment by wangyang0918** [comment]
@zentol If we define the `.out` output as what would be printed to the console(including the logging redirection), then this is indeed the expected behavior. After a careful consideration, i tend to agree with you that **NOT** introduce the stream redirection.

So i will remove the `StdOutErrRedirector` and related log4j/logback configuration. After that things become simpler. We just need to update the script and log4j/logback configuration. Given this, i will force push the new changes. Thanks a lot for your review and suggestion.

**Comment by zentol** [comment]
I'm currently waiting on more input from the ML before continuing the review.

In the mean time, I have an idea on how to solve the issue of stdout/logging going into the same file, and I'm curious whether you think it might work.
So far we tried to tackle the problem by changing _where_ the Flink process writes output to (stream redirections), or redirecting the console output (tee).
What if the Flink process wouldn't output _anything_ to the console instead writing everything to the files as usual, but we run a separate program that reads back the .out/.log files?
Basically, `exec <flink process stuff> | <file reader priting .log/.out files>`.

The finicky thing is figuring out which files to read, but we know the names when the process starts, so it should be _feasible_ at least?

**Comment by tillrohrmann** [comment]
Do we also know the file names in case of log file rotation?

The reason why I find this solution approach a bit less appealing is that it is behaviour which is harder to understand for the user (compared to a single `log4j.properties` file which contains the file and consoler appender).

**Comment by zentol** [comment]
That depends on whether they share a filename prefix, which would be reasonable though. Could be a problem at the point of rolling though as we could miss some messages 🤫

**Comment by wangyang0918** [comment]
I also prefer to use the `ConsoleAppender` in log4j. It is more straightforward and stable.

From the discussion in the ML, it seems that the only concern is about the disk consumption. So do we need to use the `RollingFileAppender` by default in `log4j-console.properties` and set the `MaxFileSize` and `MaxBackupIndex` explicitly?

@tillrohrmann @zentol WDYT?

**Comment by zentol** [comment]
Of course the `ConsoleAppender` is a better approach, there's no doubt about that. My idea was purely for the case where we _absolutely_ want to keep the log/stdout files separate.

**Comment by tillrohrmann** [comment]
Is it possible to limit the size of the `out` files? If not, then this might be an issue.

**Comment by wangyang0918** [comment]
Since we use tee to redirect the stdout to `out` file, it is hard to support rolling file and control the file size. This is also a reason for why we could consider the stream redirection. No duplication, file size control and better readability.

**Comment by tillrohrmann** [comment]
Hmm, this makes it a bit problematic. It seems as if every possible solutions has a couple of drawbacks...

**Comment by zentol** [comment]
I'd then suggest to not create a .out file.

My guess is that the lack of .log files is the greater issue anyway.

**Comment by wangyang0918** [comment]
You are right. From our internal use case, the biggest problem is `.log` could not be accessed via web dashboard. So maybe we could start with this and leave the `.out` file later until we reach a consensus or find a reasonable solution.

Moreover, i suggest to use the `RollingFileAppender` as default, set `MaxFileSize` to 100MB and `MaxBackupIndex` to 10, which means the logging file will consume 1G disk at most.

@zentol @tillrohrmann What do you think?

**Comment by zentol** [comment]
@wangyang0918 That sounds like a good default to me.

**Comment by wangyang0918** [comment]
@zentol I have updated the PR according to the discussion. We will start with `.log` file by using rolling appender. I think it could meet the most requirements and leave the open question about `.out` and `.err` file later.

File: flink-dist/src/main/flink-bin/bin/flink-console.sh [hunk 9]
```diff
@@ -58,7 +58,34 @@ esac
...
+# The lock needs to be released after use because this script is started foreground
+command -v flock >/dev/null 2>&1
+flock_exist=$?
+if [[ ${flock_exist} -eq 0 ]]; then
+    exec 200<"$FLINK_PID_DIR"
+    flock 200
+fi
+# Remove the pid file when all the processes are dead
+if [ -f "$pid" ];then
+    all_dead=0
```
Comment (zentol) [inline flink-dist/src/main/flink-bin/bin/flink-console.sh:76]: Why are we re-inventing the wheel and not use the same logic from `flink-daemon.sh`?

`[ $(wc -l < "$pid") -eq 0 ] && rm "$pid"`
or
`[ -z $(tail -n 1 "$pid") ] && rm "$pid"`

File: flink-dist/src/main/flink-bin/bin/flink-console.sh [hunk 10]
```diff
@@ -58,7 +58,34 @@ esac
...
+if [ "$FLINK_IDENT_STRING" = "" ]; then
+    FLINK_IDENT_STRING="$USER"
+fi
+
+pid=$FLINK_PID_DIR/flink-$FLINK_IDENT_STRING-$SERVICE.pid
+mkdir -p "$FLINK_PID_DIR"
+# The lock needs to be released after use because this script is started foreground
+command -v flock >/dev/null 2>&1
+flock_exist=$?
+if [[ ${flock_exist} -eq 0 ]]; then
```
Comment (zentol) [inline flink-dist/src/main/flink-bin/bin/flink-console.sh:70]:
```suggestion
if [[ $? -eq 0 ]]; then
```
stay consistent with other scripts, so it is more obvious what can be deduplicated later.

File: flink-dist/src/main/flink-bin/bin/flink-console.sh [hunk 11]
```diff
@@ -70,4 +97,10 @@ if [[ ${JAVA_VERSION} =~ ${IS_NUMBER} ]]; then
 fi
 
 echo "Starting $SERVICE as a console application on host $HOSTNAME."
+
+echo $$ >> "$pid" 2>/dev/null
```
Comment (zentol) [inline flink-dist/src/main/flink-bin/bin/flink-console.sh:101]: please add a comment for what this does

File: flink-dist/src/main/flink-bin/bin/flink-console.sh (hunk 9 above)
Comment (wangyang0918) [reply flink-dist/src/main/flink-bin/bin/flink-console.sh:76]: For `flink-daemon.sh`, it provides the `stop` and `stop-all` command so that we could remove the `pid` file there. However, in `flink-console.sh`, the script could be crashed or killed at any time. We may not have chance to remove it. So i put the clean-up logic when starting a new one. Only when all the processes are dead, we could safely delete the `pid` file.

File: flink-dist/src/main/flink-bin/bin/flink-console.sh (hunk 10 above)
Comment (wangyang0918) [reply flink-dist/src/main/flink-bin/bin/flink-console.sh:70]: The `flock_exist` will be reused later when to release the lock. So we need a variable to store it.

File: flink-dist/src/main/flink-bin/bin/flink-console.sh (hunk 11 above)
Comment (wangyang0918) [reply flink-dist/src/main/flink-bin/bin/flink-console.sh:101]: I will add a comments here.

Comment (zentol) [reply flink-dist/src/main/flink-bin/bin/flink-console.sh:102]: so `$$` returns the PID of the current process?

> and then replace the current program with specified SERVICE in the same process

With this you mean the exec call? if so, I'd omit it, because it raises more questions than it answers. (most notably due to describing behavior several lines down)

File: flink-dist/src/main/flink-bin/bin/flink-console.sh (hunk 9 above)
Comment (zentol) [reply flink-dist/src/main/flink-bin/bin/flink-console.sh:76]: Shouldn't this cleanup be done in a trap of some sort? Like on-exit remove the PID of this process from the file, and then _maybe_ delete it if empty?

File: flink-dist/src/main/flink-bin/bin/flink-console.sh [hunk 12]
```diff
@@ -70,4 +97,10 @@ if [[ ${JAVA_VERSION} =~ ${IS_NUMBER} ]]; then
...
+
+# Release the lock
```
Comment (zentol) [inline flink-dist/src/main/flink-bin/bin/flink-console.sh:104]:
```suggestion
# Release the lock, because the java process runs in the foreground and would block other processes from modifying the pid file
```

File: flink-dist/src/main/flink-bin/bin/flink-console.sh (hunk 9 above)
Comment (wangyang0918) [reply flink-dist/src/main/flink-bin/bin/flink-console.sh:76]: It is not easy to be done in a `trap` because we need to handle all the signals. The scripts also could be killed -9, then it will not have chance to clean-up. So i suggest to do this before start.

File: flink-dist/src/main/flink-bin/bin/flink-console.sh (hunk 11 above)
Comment (wangyang0918) [reply flink-dist/src/main/flink-bin/bin/flink-console.sh:102]: Yes, `$$` will return the pid of current process. For the `exec`, i will remove the comments about it.

**Comment by zentol** [comment]
@flinkbot run azure

**Comment by wangyang0918** [comment]
@zentol Thanks a lot for the `fixup`.
I have rebased the latest master and force pushed.
